src/openslide-vendor-dicom.c

```diff
@@ -0,0 +1,943 @@
+  struct associated *a = (struct associated *) _img;
+
+  DcmError *dcm_error = NULL;
+  DcmFrame *frame = dcm_filehandle_read_frame(&dcm_error,
+                                              a->file->filehandle, 
+                                              a->file->metadata,
+                                              a->file->bot, 
+                                              1);
+  if (frame == NULL) {
+    set_gerror_from_dcm_error(err, &dcm_error);
+    return false;
+  }
+
+#ifdef DEBUG
+  print_frame(frame);
+#endif /*DEBUG*/
+
+  return _openslide_jpeg_decode_buffer(dcm_frame_get_value(frame),
+                                       dcm_frame_get_length(frame),
+                                       dest,
+                                       a->base.w,
+                                       a->base.h,
+                                       err);
+}
+
+static void associated_destroy(struct _openslide_associated_image *_img) {
+  struct associated *a = (struct associated *) _img;
+  FREEF(dicom_file_destroy, a->file);
+  g_slice_free(struct associated, a);
+}
+
+static const struct _openslide_associated_image_ops dicom_associated_ops = {
+  .get_argb_data = associated_get_argb_data,
+  .destroy = associated_destroy
+};
+
+static bool read_whole_file(struct dicom_file *f) {
+  // try to read the rest of the dicom file 
+  if (!f->metadata) {
+    f->metadata = dcm_filehandle_read_metadata(NULL, f->filehandle);
+  }
+  if (!f->metadata) {
+    return false;
+  }
+
+  if (!f->bot) {
+    f->bot = dcm_filehandle_read_bot(NULL, f->filehandle, f->metadata);
+  }
+  if (!f->bot) {
+    /* Try to build the BOT instead.
+     */
+    f->bot = dcm_filehandle_build_bot(NULL, f->filehandle, f->metadata);
+  }
+  if (!f->bot) {
+    return false;
+  }
+
+  return true;
+}
+
+static struct associated *associated_new(struct dicom_file *f) {
+  // we use BOT to get the label and overview frames
+  if (!read_whole_file(f)) { 
+    return NULL;
+  }
+
+  // ImageType must be one of the combinations we accept
+  if (!is_type(f, &associated_types)) {
+    return NULL;
+  }
+
+  const char *name;
+  if (!get_tag_str(f->metadata, "ImageType", 2, &name)) {
+    return NULL;
+  }
+
+  int image_width;
+  int image_height;
+  if (!get_tag_int(f->metadata, "TotalPixelMatrixColumns", &image_width) ||
+    !get_tag_int(f->metadata, "TotalPixelMatrixRows", &image_height)) {
+    return NULL;
+  }
+
+  // set this once we know we will succeed ... this passes ownership to the
+  // level
+  struct associated *a = g_slice_new0(struct associated);
+  a->file = f;
+  a->name = name;
+  a->base.ops = &dicom_associated_ops;
+  a->base.w = image_width;
+  a->base.h = image_height;
+
+#ifdef DEBUG
+  printf("associated_new: %s\n", a->name);
+#endif /*DEBUG*/
+
+  return a;
+}
+
+static struct level *level_new(struct dicom_file *f) {
+  if (!read_whole_file(f)) { 
+    return NULL;
+  }
+
+  struct level *l = g_slice_new0(struct level);
```

Comment:
This is a constructor, so I think we need this to live outside the scope of this function. 

I allocated the `level` before I was certain construction was going to succeed since I wanted somewhere to note the file properties it has to check. I could read the properties to locals instead if that would be preferred.

Reply:
Yup, exactly.  I generally don't bother with `g_autoptr` when we're just going to infallibly set some fields and return, but if there are error paths we generally avoid manual frees.

Reply:
OK, done!